### DISCUSSION TITLE
0.5.0.0

### DISCUSSION THREAD

**Discussion by astubbs** [confluentinc/parallel-consumer#288] [General]
### Features

- feature: Poll Context object for API (#223)
  - PollContext API - provides central access to result set with various convenience methods as well as metadata about records, such as failure count
- major: Batching feature and Event system improvements
  - Batching - all API methods now support batching.
See the Options class set batch size for more information.

### Fixes and Improvements

* Event system - better CPU usage in control thread
* Concurrency stability improvements
* Update dependencies
* #247: Adopt Truth-Generator (#249)
  * Adopt https://github.com/astubbs/truth-generator[Truth Generator] for automatic generation of https://truth.dev/[Google Truth] Subjects
* Large rewrite of internal architecture for improved maintence and simplicity which fixed some corner case issues
  * refactor: Rename PartitionMonitor to PartitionStateManager (#269)
  * refactor: Queue unification (#219)
  * refactor: Partition state tracking instead of search (#218)
  * refactor: Processing Shard object
* fix: Concurrency and State improvements (#190)

### Build

* build: Lock TruthGenerator to 0.1 (#272)
* build: Deploy SNAPSHOTS to maven central snaphots repo (#265)
* build: Update Kafka to 3.1.0 (#229)
* build: Crank up Enforcer rules and turn on ossindex audit
* build: Fix logback dependency back to stable
* build: Upgrade TestContainer and CP

# Raw

## What's Changed
* build: Update Kafka to 3.1.0 by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/229
* build(deps): bump ossindex-maven-plugin from 3.1.0 to 3.2.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/244
* build(deps): bump maven-dependency-plugin from 3.2.0 to 3.3.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/245
* refactor: Partition state tracking instead of search by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/218
* build: Fix maven repetition by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/251
* #247: Adopt Truth-Generator by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/249
* build(deps): bump mycila.version from 4.2.rc2 to 4.2.rc3 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/253
* build(deps): bump maven-clean-plugin from 3.1.0 to 3.2.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/257
* build: Deploy SNAPSHOTS to maven central snaphots repo by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/265
* refactor: Queue unification by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/219
* refactor: Rename PartitionMonitor to PartitionStateManager by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/269
* build: Lock TruthGenerator to 0.1 by @astubbs in https://github.com/confluentinc/parallel-consumer/pull/272
* build(deps): bump jacoco-maven-plugin from 0.8.7 to 0.8.8 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/256
* build(deps): bump junit-pioneer from 1.6.2 to 1.7.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/260
* build(deps-dev): bump podam from 7.2.8.RELEASE to 7.2.9.RELEASE by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/261
* build(deps): bump testcontainers.version from 1.16.3 to 1.17.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/263
* build(deps): bump reactor-core from 3.4.16 to 3.4.17 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/274
* build(deps-dev): bump postgresql from 42.3.3 to 42.3.4 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/277
* build(deps): bump maven-javadoc-plugin from 3.3.2 to 3.4.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/276
* build(deps): bump mockito.version from 4.4.0 to 4.5.1 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/278
* build(deps): bump maven-site-plugin from 3.11.0 to 3.12.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/280
* build(deps): bump testcontainers.version from 1.17.0 to 1.17.1 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/279
* build(deps): bump nexus-staging-maven-plugin from 1.6.12 to 1.6.13 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/282
* build(deps): bump maven-project-info-reports-plugin from 3.2.2 to 3.3.0 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/283
* build(deps): bump vertx.version from 4.2.6 to 4.2.7 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/281
* build(deps): bump lombok from 1.18.22 to 1.18.24 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/284
* build(deps): bump wiremock from 2.32.0 to 2.33.2 by @dependabot in https://github.com/confluentinc/parallel-consumer/pull/285

**Full Changelog**: https://github.com/confluentinc/parallel-consumer/compare/0.4.0.1...0.5.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/confluentinc/parallel-consumer/releases/tag/0.5.0.0'>0.5.0.0</a>.</em>

**Comment by JorgenRingen** [comment]
Awesome and congratulations! 🥳

**Reply by astubbs** [reply]
haha, thank you! It was a big rewrite of the most poorly written parts of the internals :) much simpler now, and adding the most of the remaining things in the 1.0 train should be much simpler now... Also will be much easier for the the community to contribute to.
